### PR TITLE
Properly reflect control flow for lifted Boolean AndAlso/OrElse operators in control flow graph.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Can any of these types be null for error case? For example, if one of the operands is `null` literal.

File: src/Compilers/Core/Portable/Symbols/ITypeSymbol.cs
Comment (mavasani) [inline src/Compilers/Core/Portable/Symbols/ITypeSymbol.cs]: >. [](start = 23, length = 1)

Should we use conditional access here to avoid null checks in callers for error cases?

**Review by mavasani — APPROVED** [review]
:shipit:

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Can any of these types be null for error case? For example, if one of the operands is null literal.

I do not think it is possible given the current implementation of compilers.

---
In reply to: [189607822](https://github.com/dotnet/roslyn/pull/26994#discussion_r189607822) [](ancestors = 189607822)

File: src/Compilers/Core/Portable/Symbols/ITypeSymbol.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Symbols/ITypeSymbol.cs]: > Should we use conditional access here to avoid null checks in callers for error cases?

So far we didn't have to make the null checks at the call sites. However, in next PRs I'll add conditional access to these helpers to make CFG builder code and future usages more robust. 

---
In reply to: [189610315](https://github.com/dotnet/roslyn/pull/26994#discussion_r189610315) [](ancestors = 189610315)

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler Please review

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Might be good to add a note that, by the VB spec, if the first operand is Nothing, we evaluate the second operand anyway, but the result is guaranteed to be Nothing.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Prototype comment isn't addressed?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Just so the next person isn't as confused as I was on initial read.

---
In reply to: [189756028](https://github.com/dotnet/roslyn/pull/26994#discussion_r189756028) [](ancestors = 189756028)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Prototype comment isn't addressed?

**Comment by 333fred** [comment]
Done review pass (iteration 1). Want confirmation on the prototype comments before approval.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Prototype comment isn't addressed?

No

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Prototype comment isn't addressed?

No

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Might be good to add a note that, by the VB spec, if the first operand is Nothing, we evaluate the second operand anyway, but the result is guaranteed to be Nothing.

This is what three-value logic mean

---
In reply to: [189756507](https://github.com/dotnet/roslyn/pull/26994#discussion_r189756507) [](ancestors = 189756507,189756028)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Prototype comment isn't addressed?

To clarify, it is not addressed, but the comment is moved into the helper that is supposed to deal with that. 

---
In reply to: [189756446](https://github.com/dotnet/roslyn/pull/26994#discussion_r189756446) [](ancestors = 189756446)

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler Please review

**Review by 333fred — APPROVED** [review]
LGTM (iteration 1)

**Comment by AlekseyTs** [comment]
@jcouv, @jaredpar, @dotnet/roslyn-compiler Please review, need second review

**Comment by jcouv** [comment]
Sorry for the delay. I'm looking.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jcouv) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: That confused me as well (+1 on adding a comment please).
I looked for the lowering logic for lifted Or on nullable bools (`RewriteLiftedBooleanBinaryOperator` I think) and there I saw some checks for `left.HasValue` and `right.HasValue` in the pattern. I am looking in the wrong place?

**Review by jcouv — APPROVED** [review]
LGTM

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > I am looking in the wrong place?

Probably the right place. It is not a goal to duplicate the lowering as compiler does, the goal is to accurately reflect the semantics, the order of evaluation and conditional nature of evaluation. 

---
In reply to: [190112981](https://github.com/dotnet/roslyn/pull/26994#discussion_r190112981) [](ancestors = 190112981)

**Pull request merged**
